### PR TITLE
PPS: association cut fix

### PR DESCRIPTION
**Pull request by jan-kaspar**
#### PR description:

This is a bugfix PR to address the issues raised in:
  * https://github.com/cms-sw/cmssw/issues/35928
  * https://github.com/cms-sw/cmssw/issues/35936

The proposal is based mainly on these suggestions:
  * https://github.com/cms-sw/cmssw/issues/35928#issuecomment-956208463
  * https://github.com/cms-sw/cmssw/issues/35936#issuecomment-956242043

#### PR validation:

The plots below compare results before (blue) and after this PR (red dashed):
  * [dirsim_cmp.pdf](https://github.com/cms-sw/cmssw/files/7453777/dirsim_cmp.pdf): for PPS direct simulation (+ reconstruction), Run2 and Run3
  * [reco_cmp.pdf](https://github.com/cms-sw/cmssw/files/7453779/reco_cmp.pdf): for PPS reconstruction, Run2 data

No differences observed as expected.

### CODE REVIEW

**Comment by malbouis** [comment]
please test

**Comment by malbouis** [comment]
type bug-fix

**Comment by malbouis** [comment]
please test with #35921

**Comment by qliphy** [comment]
enable threading

**Comment by qliphy** [comment]
test parameters:

- workflow=136.796,136.816,136.8642,136.8861,23234.103
- pull_request= #35291

**Comment by qliphy** [comment]
please test

**Comment by qliphy** [comment]
please abort

**Comment by qliphy** [comment]
enable threading

**Comment by qliphy** [comment]
test parameters:
-     workflow=136.796,136.816,136.8642,136.8861,23234.103
-     pull_request=  35921

**Comment by qliphy** [comment]
please test

**Comment by perrotta** [comment]
urgent

**Comment by malbouis** [comment]
@jan-kaspar there seems to be a .at() left in L.78 of of PPSAssociationCuts.h. I think it should replaced with .find()

File: CondFormats/PPSObjects/interface/PPSAssociationCuts.h [hunk 1]
```diff
@@ -70,6 +69,12 @@ class PPSAssociationCuts {
 
   ~PPSAssociationCuts() {}
 
+  // builds run-time data members, useful e.g. after loading data from DB
+  void initialize() {
+    for (auto &p : association_cuts_)
+      p.second.buildFunctions();
+  }
+
   const CutsPerArm &getAssociationCuts(const int sector) const { return association_cuts_.at(sector); }
```
Comment (tvami) [inline CondFormats/PPSObjects/interface/PPSAssociationCuts.h:78]: this is the line @malbouis meant

Comment (jan-kaspar) [reply CondFormats/PPSObjects/interface/PPSAssociationCuts.h:78]: Thanks! I see now. At the moment of opening the PR, there was only this answer: https://github.com/cms-sw/cmssw/issues/35936#issuecomment-956242043, which I interpret that `at` is OK for a map. Later, another answer was provided: https://github.com/cms-sw/cmssw/issues/35936#issuecomment-956255463, which seems to be in disagreement with the first one. I guess I need a decision which hint to follow. If there is no clear guideline, keeping the code as is saves my time.

Comment (malbouis) [reply CondFormats/PPSObjects/interface/PPSAssociationCuts.h:78]: @makortel @VinInn , is it ok to keep the .at() in this case?

Comment (tvami) [reply CondFormats/PPSObjects/interface/PPSAssociationCuts.h:78]: Hi @jan-kaspar my interpretation is that Matti is saying `at` is OK if we are OK with throwing an exception, while Vincenzo is saying it is not OK to throw the exception, thus `at` is also not OK. I think `find` should be used.

Comment (VinInn) [reply CondFormats/PPSObjects/interface/PPSAssociationCuts.h:78]: The real question is: does one really need a ```map```?
Is not a trivial array enough?  (what values of "arm" should be supported?)

Comment (makortel) [reply CondFormats/PPSObjects/interface/PPSAssociationCuts.h:78]: What would be the circumstances if an element is not found? I'm wondering mostly between `sector` comes from data (directly or via some processing) and can go out of bounds occasionally, and `sector` originally comes from configuration and out of bounds values would be logic errors.

What would the calling code do if an element is not found?

Comment (jan-kaspar) [reply CondFormats/PPSObjects/interface/PPSAssociationCuts.h:78]: The `sector` refers to the LHC sectors around IP5: these are two and will never change. I agree that array/vector would have been a better design. The code has been made by a student and I didn't catch this - sorry for that. Now that the class is already used in CondDB, I don't want to change it anymore. On the other hand, doing the search (by map::find) on a map with two elements is not that costly, so I believe we are having a sort of academic discussion. I will commit the change to map::find soon.

Comment (jan-kaspar) [reply CondFormats/PPSObjects/interface/PPSAssociationCuts.h:78]: For the moment there is no guarantee in the code that the DB payload will contain data for both LHC sectors. We assume that the person who uploads the conditions to DB does this well. Let me know if you wish me to put some "protection" to the code.

**Comment by tvami** [comment]
@cmsbuild , please test

File: CondFormats/PPSObjects/interface/PPSAssociationCuts.h [hunk 2]
```diff
@@ -70,7 +69,13 @@ class PPSAssociationCuts {
...
   ~PPSAssociationCuts() {}
 
-  const CutsPerArm &getAssociationCuts(const int sector) const { return association_cuts_.at(sector); }
+  // builds run-time data members, useful e.g. after loading data from DB
+  void initialize() {
+    for (auto &p : association_cuts_)
+      p.second.buildFunctions();
+  }
+
+  const CutsPerArm &getAssociationCuts(const int sector) const { return association_cuts_.find(sector)->second; }
```
Comment (makortel) [inline CondFormats/PPSObjects/interface/PPSAssociationCuts.h:78]: I just want to point out that with this approach a incorrect payload or `sector` value lead to undefined behavior (crashes at best, silent garbage output at worst).

Comment (VinInn) [reply CondFormats/PPSObjects/interface/PPSAssociationCuts.h:78]: doing a check and throw (ie abort) at inizialize time is ok.
What is NOT ok is to check validity at each single access

**Comment by tvami** [comment]
+1
 * multi-threaded tests pass now
 * follow-up to the review happened according to the feedback

**Comment by qliphy** [comment]
+1

**Pull request merged**